### PR TITLE
add ViT classification

### PR DESCRIPTION
**Pull request by GuoQuanhao**
为CIFAR-100增加ViT分类tutorial

之前的pr：https://github.com/PaddlePaddle/docs/pull/5513

### CODE REVIEW

File: docs/practices/cv/image_classification_ViT.ipynb [hunk 1]
```diff
@@ -0,0 +1,844 @@
...
+    "plot_num_images(25, cifar100_test)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 模型组网\n",
+    "\n",
+    "ViT模型组网部分包含图像切片Patches，Implement multilayer perceptron (MLP)，多头自注意力机制（MultiHeadSelfAttention）以及TransformerEncoder"
```
Comment (TingquanGao) [inline docs/practices/cv/image_classification_ViT.ipynb]: Implement multilayer perceptron -> Multilayer perceptron

**Review by TingquanGao — COMMENTED** [review]
再加一下标题序号吧：
一、简要介绍
二、环境配置
三、加载数据集
四、模型组网
4.1 Patches

等等，如上这样。

**Comment by TingquanGao** [comment]
基于这个PR改就行，不用提新的pr。

**Comment by GuoQuanhao** [comment]
已修改

File: docs/practices/cv/image_classification_ViT.ipynb [hunk 2]
```diff
@@ -0,0 +1,844 @@
...
+    "        x = x.reshape(new_shape)\n",
+    "        x = x.transpose([0, 2, 1, 3])\n",
+    "        return x\n",
+    "\n",
+    "    def forward(self, x):\n",
+    "        b, n, f = x.shape\n",
+    "        qkv = self.qkv(x).chunk(3, -1)\n",
+    "        q, k, v = map(self.transpose_multi_head, qkv)\n",
+    "        score = F.softmax(paddle.einsum(\"bhif, bhjf->bhij\", q, k)/self.sqrt_d, axis=-1)\n",
+    "        attn = paddle.einsum(\"bhij, bhjf->bihf\", score, v)\n",
```
Comment (TingquanGao) [inline docs/practices/cv/image_classification_ViT.ipynb]: 这里官方的ViT是需要对attn做dropout吧？为什么这里去掉了一个dropout。

Comment (GuoQuanhao) [reply docs/practices/cv/image_classification_ViT.ipynb]: out = self.dropout(self.out(attn.flatten(2)))

Comment (GuoQuanhao) [reply docs/practices/cv/image_classification_ViT.ipynb]: 好像是差了个dropout，我改下

**Comment by GuoQuanhao** [comment]
为MSA增加了dropout，删除了MLP中的一个act

**Comment by sunzhongkai588** [comment]
CI没过，请先用 pre-commit 查看是否有符合规范

**Review by TingquanGao — APPROVED** [review]

File: docs/practices/cv/image_classification_ViT.ipynb [hunk 3]
```diff
@@ -0,0 +1,847 @@
...
+     "output_type": "stream",
+     "text": [
+      "2.4.0\n"
+     ]
+    },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "/opt/conda/envs/python35-paddle120-env/lib/python3.7/site-packages/matplotlib/__init__.py:107: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working\n",
```
Comment (sunzhongkai588) [inline docs/practices/cv/image_classification_ViT.ipynb]: image
这部分warning请检查一下，尽量不要展示或暴露

File: docs/practices/cv/image_classification_ViT.ipynb [hunk 4]
```diff
@@ -0,0 +1,847 @@
...
+    "摘要: 本示例教程演示如何在CIFAR-100数据集上用ViT进行图像分类。"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 一、简要介绍\n",
+    "![](https://ai-studio-static-online.cdn.bcebos.com/063de5fc4b814ad6bfd976320f7c7cc9ca06b38744b1458189ae3a7356f3ffd4)\n",
+    "ViT尽量保持NLP中Transformer结构，做尽量小的结构或代码调整，主要目的是证明NLP中的Transformer能够很好的迁移到视觉任务上\n",
```
Comment (sunzhongkai588) [inline docs/practices/cv/image_classification_ViT.ipynb]: 在这句话前，先用几句话简短介绍下vit是什么，并附上论文链接 https://arxiv.org/abs/2010.11929 （因为下文有提到“..作者..”）

File: docs/practices/cv/image_classification_ViT.ipynb [hunk 5]
```diff
@@ -0,0 +1,847 @@
...
+    "\n",
+    "- 使用图像切片的方式将图片转变成一组Patch得到一个序列；\n",
```
Comment (sunzhongkai588) [inline docs/practices/cv/image_classification_ViT.ipynb:27]: 在这些操作前，一句话简要说明下这些步骤是做什么的，如 `以下步骤简要阐述了....`
image

File: docs/practices/cv/image_classification_ViT.ipynb [hunk 6]
```diff
@@ -0,0 +1,847 @@
...
+    "\n",
+    "print(paddle.__version__)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 三、加载数据集\n",
+    "本案例将会使用飞桨提供的API完成数据集的下载并为后续的训练任务准备好数据迭代器。CIFAR-100数据集有100个类。每个类有600张大小为$32\\times 32$的彩色图像，其中500张作为训练集，100张作为测试集"
```
Comment (sunzhongkai588) [inline docs/practices/cv/image_classification_ViT.ipynb]: 标点符号请用准确～

`本案例将会使用飞桨提供的API完成数据集的下载，并为后续的训练任务准备好数据迭代器。CIFAR-100数据集有100个类，每个类有600张大小为`........(省略中间内容）...`100张作为测试集。`

File: docs/practices/cv/image_classification_ViT.ipynb [hunk 7]
```diff
@@ -0,0 +1,847 @@
...
+    "\n",
+    "print('训练集数量:', len(cifar100_train), '训练集图像尺寸', cifar100_train[0][0].shape)\n",
+    "print('测试集数量:', len(cifar100_test), '测试集图像尺寸', cifar100_test[0][0].shape)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "显示训练和测试的图像，其中训练图像包含数据增强"
```
Comment (sunzhongkai588) [inline docs/practices/cv/image_classification_ViT.ipynb]: 请写完整，如：

`显示训练集和测试集的图像，其中训练集图像包含数据增强后的图像`

File: docs/practices/cv/image_classification_ViT.ipynb (hunk 6 above)
Comment (sunzhongkai588) [inline docs/practices/cv/image_classification_ViT.ipynb]: 也加一段话，阐述下面的代码是在做什么

image

File: docs/practices/cv/image_classification_ViT.ipynb (hunk 1 above)
Comment (sunzhongkai588) [reply docs/practices/cv/image_classification_ViT.ipynb]: @GuoQuanhao    请把该句尽量写的工整，如
`ViT模型组网部分包含图像切片（Patches），多层感知机（MLP），多头自注意力机制（MultiHeadSelfAttention）以及Transformer编码器（Transforme Encoder）`

File: docs/practices/cv/image_classification_ViT.ipynb [hunk 8]
```diff
@@ -0,0 +1,847 @@
...
+   ],
+   "source": [
+    "plot_num_images(25, cifar100_train)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "测试集图片可视化"
```
Comment (sunzhongkai588) [inline docs/practices/cv/image_classification_ViT.ipynb]: 请check一下，没有展示结果。
image

File: docs/practices/cv/image_classification_ViT.ipynb [hunk 9]
```diff
@@ -0,0 +1,847 @@
...
+    "    def forward(self, images):\n",
+    "        patches = F.unfold(images, self.patch_size, self.patch_size)\n",
+    "        return patches.transpose([0,2,1])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "可视化产生的image patches"
```
Comment (sunzhongkai588) [inline docs/practices/cv/image_classification_ViT.ipynb:322]: 似乎也没有输出图片？

image

File: docs/practices/cv/image_classification_ViT.ipynb [hunk 10]
```diff
@@ -0,0 +1,847 @@
...
+    "    plt.axis(\"off\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 4.2 MLP\n",
+    "\n",
+    "多层感知机。包含线性层，激活层(GELU)，DropOut层，将输入扩增指定维度，再缩减回去，MLP不改变输入输出维度"
```
Comment (sunzhongkai588) [inline docs/practices/cv/image_classification_ViT.ipynb]: 注意语句、标点等请书写规范，如：

多层感知机包含线性层、激活层(GELU)、DropOut层。线性层将输入扩增指定维度，再缩减回去，MLP不改变输入输出维度

File: docs/practices/cv/image_classification_ViT.ipynb [hunk 11]
```diff
@@ -0,0 +1,847 @@
...
+    "        return x"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 4.3 Multi-Head Self Attention(MSA)\n",
+    "\n",
+    "多头自注意力机制。每个token的维度是768，假如设置12个头，每个头的维度为$768/12=64$，得到$Q,K,V=197×64$维度。最后将12个头的输出拼接起来最后又得到$197×768$维度。论文发现自注意力在网络刚开始时就能注意到全局上的信息，而不像CNN一样在浅层部分感受野非常小，在深层部分，自注意力距离非常可以非常远，能够学到带有语义性的概念，而不是靠临近的像素点去判断\n",
```
Comment (sunzhongkai588) [inline docs/practices/cv/image_classification_ViT.ipynb]: `论文发现自注意力在网络刚开始时就能注意到全局上的信息，而不像CNN一样在浅层部分感受野非常小，在深层部分，自注意力距离非常可以非常远，能够学到带有语义性的概念，而不是靠临近的像素点去判断` 语句不通顺，请斟酌并修改

File: docs/practices/cv/image_classification_ViT.ipynb [hunk 12]
```diff
@@ -0,0 +1,847 @@
...
+    "<center class=\"half\"><img src=https://ai-studio-static-online.cdn.bcebos.com/8129768a52b04f898d9115ca39a384e8fea7e26dd9f94cf5bd2411a1ad0553c1 width=\"400\"/>&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp<img src=https://ai-studio-static-online.cdn.bcebos.com/d61d6660198243a1afc4027b0ac3a7446864458b8fc149ecb8861ffd6fa1dd70 width=\"200\"/> </center>"
```
Comment (sunzhongkai588) [inline docs/practices/cv/image_classification_ViT.ipynb]: 该部分check一下
image

File: docs/practices/cv/image_classification_ViT.ipynb [hunk 13]
```diff
@@ -0,0 +1,847 @@
...
+    "        out = self.fc(out)\n",
+    "        return out"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 五、Slover构建\n",
+    "为了取得更好的效果，构建LabelSmoothingCrossEntropyLoss作为损失函数，并采用带Warmup的Cosine学习率衰减方式"
```
Comment (sunzhongkai588) [inline docs/practices/cv/image_classification_ViT.ipynb]: `带Warmup的Cosine学习率衰减方式`建议附上相应api文档的链接

File: docs/practices/cv/image_classification_ViT.ipynb [hunk 14]
```diff
@@ -0,0 +1,847 @@
...
+   },
+   "outputs": [],
+   "source": [
+    "schedulerTest = get_scheduler(epochs=100, warmup_epochs=5, learning_rate=1e-3)\n",
+    "lr = []\n",
+    "for epoch in range(100):\n",
+    "    lr.append(schedulerTest.get_lr())\n",
+    "    schedulerTest.step()\n",
+    "plt.plot(lr)\n",
+    "plt.show()"
```
Comment (sunzhongkai588) [inline docs/practices/cv/image_classification_ViT.ipynb:650]: 同理，输出没有显示

image

File: docs/practices/cv/image_classification_ViT.ipynb [hunk 15]
```diff
@@ -0,0 +1,847 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 六、模型训练\n",
+    "\n",
+    "在训练之前可以通过paddle.summary查验模型"
```
Comment (sunzhongkai588) [inline docs/practices/cv/image_classification_ViT.ipynb]: - paddle.summary 附上相应api的文档链接
- 同理，没有显示结果。（没有显示输出结果是个本pr存在的共性问题，麻烦校对一下～）
image

File: docs/practices/cv/image_classification_ViT.ipynb [hunk 16]
```diff
@@ -0,0 +1,847 @@
...
+   "source": [
+    "start = time.time()\n",
+    "for epoch in range(EPOCHS):\n",
+    "    train_epoch(Model)\n",
+    "    validation(Model)\n",
+    "    if (epoch + 1) % 50 == 0:\n",
+    "        paddle.save(Model.state_dict(), str(epoch + 1) + '.pdparams')\n",
+    "paddle.save(Model.state_dict(), 'finished.pdparams')\n",
+    "end = time.time()\n",
+    "print('Training Cost ', (end-start) / 60, 'minutes')\n"
```
Comment (sunzhongkai588) [inline docs/practices/cv/image_classification_ViT.ipynb]: 这部分是没有训练完吗？
image

File: docs/practices/cv/image_classification_ViT.ipynb [hunk 17]
```diff
@@ -0,0 +1,847 @@
...
+    "jupyter": {
+     "outputs_hidden": false
+    }
+   },
+   "outputs": [],
+   "source": [
+    "Model.eval()\n",
+    "data = cifar100_test[np.random.randint(0, len(cifar100_test))]\n",
+    "pred = Model(paddle.to_tensor(data[0]).unsqueeze(0)).argmax(-1).item()\n",
+    "print('预测结果为:', pred, '标签为:', data[1].item())"
```
Comment (sunzhongkai588) [inline docs/practices/cv/image_classification_ViT.ipynb:1061]: 同理，结果没有展示
image

**Review by sunzhongkai588 — COMMENTED** [review]
辛苦逐个修改一下相应的问题～

File: docs/practices/cv/image_classification_ViT.ipynb (hunk 3 above)
Comment (GuoQuanhao) [reply docs/practices/cv/image_classification_ViT.ipynb]: 更新到了最新的2.4.1稳定版，并消除了警告

File: docs/practices/cv/image_classification_ViT.ipynb (hunk 4 above)
Comment (GuoQuanhao) [reply docs/practices/cv/image_classification_ViT.ipynb]: 增加了ViT的介绍，并加粗了V,i,T。阐述了与NLP之间的联系，增加了论文链接

File: docs/practices/cv/image_classification_ViT.ipynb (hunk 5 above)
Comment (GuoQuanhao) [reply docs/practices/cv/image_classification_ViT.ipynb:27]: 已修改

File: docs/practices/cv/image_classification_ViT.ipynb (hunk 6 above)
Comment (GuoQuanhao) [reply docs/practices/cv/image_classification_ViT.ipynb]: 已修改

File: docs/practices/cv/image_classification_ViT.ipynb (hunk 7 above)
Comment (GuoQuanhao) [reply docs/practices/cv/image_classification_ViT.ipynb]: 已修改

**Comment by GuoQuanhao** [comment]
已修正所有问题

**Comment by sunzhongkai588** [comment]
> 已修正所有问题

目前还有两个小问题，辛苦修复一下～

- 简要说明这两个代码块的用处，如前者是用于`定义超参数、优化器...`
image

- 是否可以缩短训练epoch的输出，如20个epoch输出一下
image

**Review by sunzhongkai588 — APPROVED** [review]
LGTM

**Review by jzhang533 — COMMENTED** [review]
很赞的文章。有两个小建议。

- 简要介绍那节里，用到的两张图片，请把他们也放到PR里，可以存放到`docs/practices/cv/sample_img`目录下，在代码仓库里维护起来。
- 简要介绍那节里，请将ViT的原始论文的出处写上吧。

**Comment by GuoQuanhao** [comment]
- 已增加论文出处
- 使用的图片已增入sample_img
- 图片路径从网络调整为仓库内

**Review by jzhang533 — APPROVED** [review]
LGTM
